src/main/scala/net/snowflake/spark/snowflake/TelemetryReporter.scala

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright 2015-2020 Snowflake Computing
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.snowflake.spark.snowflake
+
+private[snowflake] object TelemetryReporter {
+
+  private var driverTelemetryReporter: TelemetryReporter = new NoopTelemetryReporter()
+
+  private[snowflake] def setDriverTelemetryReporter(tr: TelemetryReporter): Unit = {
+    driverTelemetryReporter = tr
+  }
+
+  private[snowflake] def resetDriverTelemetryReporter(): Unit = {
+    driverTelemetryReporter = new NoopTelemetryReporter()
+  }
+
+  private val executorTelemetryReporters = new ThreadLocal[TelemetryReporter]() {
```

Comment:
Reporter, no s

Reply:
The telemetry stuff is already "thread local" why is this also a thread local ?

Reply:
My current idea is all required information is set into the ExecutorTelemetryReporter instance. Each thread has its own instance.

After careful consideration, I think it could work with singleton. What we need to do are:
1. Get TaskContext to extract task info when logging.
2. Multiple threads may use the same instance to send message, need synchronization to send logging.

I will update, since this proposal sounds simpler.